File: solution/3200-3299/3260.Find the Largest Palindrome Divisible by K/Solution.java
class Solution {
    StringBuilder sb = new StringBuilder();
    public String largestPalindrome(int n, int k) {
        if (k == 1 || k == 3 || k == 9) {
            return case139(n);
        }
        if (k == 2) {
            return case2(n);
        }
        if (k == 4) {
            return case4(n);
        }
        if (k == 5) {
            return case5(n);
        }
        if (k == 6) {
            return case6(n);
        }
        if (k == 8) {
            return case8(n);
        }
        return case7(n);
    }

    public String case139(int n) {
        while (n > 0) {
            n--;
            sb.append('9');
        }
        return sb.toString();
    }

    public String case2(int n) {
        if (n == 1)
            return "8";
        if (n == 2)
            return "88";
        return "8" + case139(n - 2) + "8";
    }

    public String case4(int n) {
        if (n == 1)
            return "8";
        if (n == 2)
            return "88";
        if (n == 3)
            return "888";
        if (n == 4)
            return "8888";
        return "88" + case139(n - 4) + "88";
    }

    public String case8(int n) {
        if (n <= 6) {
            while (n > 0) {
                sb.append("8");
                n--;
            }
            return sb.toString();
        }
        return "888" + case139(n - 6) + "888";
    }

    public String case5(int n) {
        if (n == 1)
            return "5";
        if (n == 2)
            return "55";
        return "5" + case139(n - 2) + "5";
    }

    public String case6(int n) {
        if (n == 1)
            return "6";
        if (n == 2)
            return "66";
        String s = case2(n);
        if (n % 2 == 0) {
            return s.substring(0, n / 2 - 1) + "77" + s.substring(n / 2 + 1, n);
        }
        return s.substring(0, n / 2) + "8" + s.substring(n / 2 + 1, n);
    }

    public String case7(int n) {
        if (n == 1)
            return "7";
        if (n == 2)
            return "77";
        String s = case139(n);
        if (n % 2 == 0) {
            for (int i = 9; i >= 0; i--) {
                s = s.substring(0, n / 2 - 1) + i + "" + i + s.substring(n / 2 + 1, n);
                int remainder = 0;
                for (int j = 0; j < n; j++)
                    remainder = (remainder * 10 + s.charAt(j) - '0') % 7;
                if (remainder == 0)
                    return s;
            }
            return s;
        }
        for (int i = 9; i >= 0; i--) {
            s = s.substring(0, n / 2) + i + s.substring(n / 2 + 1, n);
            int remainder = 0;
            for (int j = 0; j < n; j++)
                remainder = (remainder * 10 + s.charAt(j) - '0') % 7;
            if (remainder == 0)
                return s;
        }
        return s;
    }
}

File: solution/3200-3299/3260.Find the Largest Palindrome Divisible by K/Solution.py
class Solution:
    def is_divisible_by_7(self,number_str):
        remainder = 0
        for char in number_str:
            digit = int(char)
            remainder = (remainder * 10 + digit) % 7
        return remainder == 0
    def largestPalindrome(self, n: int, k: int) -> str:
        if k == 1: return "9" * n
        if k == 2:
            if n == 1:
                return "8"
            if n == 2:
                return "88"
            return "8" + "9"*(n-2) + "8"
        if k == 3: return "9" * n
        if k == 4:
            if n <= 4:
                return "8" * n
            return "88" + (n - 4)*"9" + "88"
        if k == 5:
            if n == 1:
                return "5"
            if n == 2:
                return "55"
            return "5" + (n - 2)*"9" + "5"
        if k == 6:
            if n <= 2:
                return "6" * n
            if n == 3:
                return "888"
            if n % 2 == 1:
                return "8" + ((n - 3)//2)*"9" + "8" + ((n - 3)//2)*"9" + "8"
            if n % 2 == 0:
                return "8" + ((n - 3) // 2) * "9" + "77" + ((n - 3) // 2) * "9" + "8"
        if k == 7:
            if n <= 2:
                return "7" * n
            if n % 2 == 1:
                nines = n // 2
                half = nines * "9"
                for i in range(9, -1, -1):
                    X = half + str(i) + half
                    if self.is_divisible_by_7(X): return X
            if n % 2 == 0:
                nines = (n // 2) - 1
                half = nines * "9"
                for i in range(9, -1, -1):
                    X = half + str(i)*2 + half
                    if self.is_divisible_by_7(X): return X
        if k == 8:
            if n <= 6: return "8" * n
            return "888" + (n - 6) * "9" + "888"
        if k == 9: return "9" *n
